File: server/internal/recording/snapshot_refs_test.go
package recording

import (
	"encoding/json"
	"errors"
	"testing"
)

func refRecordingPayload() map[string]any {
	// Mirrors the wire shape the extension emits: a full initial snapshot
	// followed by two reference snapshots and a changed final snapshot.
	raw := `{
		"version": "2.0.0",
		"snapshots": [
			{
				"timestamp": 1, "url": "https://example.com/", "selectorMap": {"1": {"index": 1, "tagName": "a", "selector": "a.next"}},
				"phase": "initial", "sequence": 0, "actionIndex": 0,
				"capture": {"status": "complete", "nodeCount": 10, "frames": []},
				"domTree": {"type": "element", "tagName": "main", "children": [{"type": "text", "text": "Example"}]}
			},
			{
				"timestamp": 2, "url": "https://example.com/", "selectorMap": {},
				"phase": "before-action", "sequence": 1, "actionIndex": 0, "ref": 0
			},
			{
				"timestamp": 3, "url": "https://example.com/", "selectorMap": {},
				"phase": "before-action", "sequence": 2, "actionIndex": 1, "ref": 0
			},
			{
				"timestamp": 4, "url": "https://example.com/changed", "selectorMap": {"2": {"index": 2, "tagName": "b", "selector": "b.done"}},
				"phase": "final", "sequence": 3, "actionIndex": 2,
				"capture": {"status": "complete", "nodeCount": 12, "frames": []},
				"domTree": {"type": "element", "tagName": "main", "children": [{"type": "text", "text": "Changed"}]}
			}
		]
	}`
	var payload map[string]any
	if err := json.Unmarshal([]byte(raw), &payload); err != nil {
		panic(err)
	}
	return payload
}

func snapshotsOf(t *testing.T, payload map[string]any) []map[string]any {
	t.Helper()
	rawList := payload["snapshots"].([]any)
	out := make([]map[string]any, 0, len(rawList))
	for _, raw := range rawList {
		out = append(out, raw.(map[string]any))
	}
	return out
}

func TestExpandSnapshotReferencesResolvesIdenticalAdjacentSnapshots(t *testing.T) {
	payload := refRecordingPayload()
	if err := ExpandSnapshotReferences(payload); err != nil {
		t.Fatalf("expand: %v", err)
	}
	snaps := snapshotsOf(t, payload)
	if len(snaps) != 4 {
		t.Fatalf("snapshot count changed: %d", len(snaps))
	}
	initial, refA, refB, final := snaps[0], snaps[1], snaps[2], snaps[3]
	for name, snap := range map[string]map[string]any{"refA": refA, "refB": refB} {
		if _, still := snap["ref"]; still {
			t.Fatalf("%s kept its ref pointer", name)
		}
		if snap["domTree"] == nil || snap["selectorMap"] == nil || snap["capture"] == nil {
			t.Fatalf("%s missing expanded content fields", name)
		}
		if got, want := marshal(t, snap["domTree"]), marshal(t, initial["domTree"]); got != want {
			t.Fatalf("%s domTree does not equal referenced content:\n got %s\nwant %s", name, got, want)
		}
		if got, want := marshal(t, snap["selectorMap"]), marshal(t, initial["selectorMap"]); got != want {
			t.Fatalf("%s selectorMap does not equal referenced content", name)
		}
		if got, want := marshal(t, snap["capture"]), marshal(t, initial["capture"]); got != want {
			t.Fatalf("%s capture does not equal referenced content", name)
		}
		// Positional identity stays per-snapshot.
		if snap["phase"] != "before-action" || snap["timestamp"] == initial["timestamp"] {
			t.Fatalf("%s lost its positional identity: %v", name, snap)
		}
	}
	if final["domTree"] == nil || final["ref"] != nil {
		t.Fatalf("changed final snapshot must remain a full snapshot: %v", final)
	}
	if marshal(t, final["domTree"]) == marshal(t, initial["domTree"]) {
		t.Fatalf("final snapshot content must differ from the referenced content")
	}
}

func TestExpandSnapshotReferencesExpandsBeforeSanitizeContract(t *testing.T) {
	payload := refRecordingPayload()
	expanded, err := json.Marshal(payload)
	if err != nil {
		t.Fatal(err)
	}
	// The expanded recording must be shape-identical to a recording that
	// never used references: dropping refs keeps selector evidence and the
	// delta archive working without any downstream awareness.
	if err := ExpandSnapshotReferences(payload); err != nil {
		t.Fatalf("expand: %v", err)
	}
	var round map[string]any
	if err := json.Unmarshal(expanded, &round); err != nil {
		t.Fatal(err)
	}
	_ = round
	sanitized, _ := Sanitize(payload)
	snaps, ok := sanitized["snapshots"].([]any)
	if !ok || len(snaps) != 4 {
		t.Fatalf("sanitized snapshots missing after expansion: %v", sanitized["snapshots"])
	}
	for index, raw := range snaps {
		snap := raw.(map[string]any)
		if snap["domTree"] == nil {
			t.Fatalf("sanitized snapshot %d lost domTree", index)
		}
	}
}

func TestExpandSnapshotReferencesFailsClosed(t *testing.T) {
	cases := map[string]func(payload map[string]any){
		"unknown target": func(p map[string]any) {
			p["snapshots"].([]any)[1].(map[string]any)["ref"] = float64(99)
		},
		"forward target": func(p map[string]any) {
			// Point the first ref at a LATER sequence (the final snapshot).
			p["snapshots"].([]any)[1].(map[string]any)["ref"] = float64(3)
		},
		"target without content": func(p map[string]any) {
			list := p["snapshots"].([]any)
			list[1].(map[string]any)["ref"] = float64(1) // names itself (a ref)
		},
		"malformed pointer": func(p map[string]any) {
			p["snapshots"].([]any)[1].(map[string]any)["ref"] = "0"
		},
		"negative pointer": func(p map[string]any) {
			p["snapshots"].([]any)[1].(map[string]any)["ref"] = float64(-1)
		},
	}
	for name, mutate := range cases {
		t.Run(name, func(t *testing.T) {
			payload := refRecordingPayload()
			mutate(payload)
			err := ExpandSnapshotReferences(payload)
			if !errors.Is(err, ErrInvalidSnapshotReference) {
				t.Fatalf("expected ErrInvalidSnapshotReference, got %v", err)
			}
		})
	}
}

func TestExpandSnapshotReferencesNoopForLegacyPayloads(t *testing.T) {
	payload := map[string]any{"version": "1.0.0", "snapshots": []any{
		map[string]any{"timestamp": 1.0, "url": "https://example.com/", "selectorMap": map[string]any{}},
	}}
	if err := ExpandSnapshotReferences(payload); err != nil {
		t.Fatalf("legacy payload rejected: %v", err)
	}
	if err := ExpandSnapshotReferences(nil); err != nil {
		t.Fatalf("nil payload rejected: %v", err)
	}
	if err := ExpandSnapshotReferences(map[string]any{"version": "2.0.0"}); err != nil {
		t.Fatalf("snapshot-less payload rejected: %v", err)
	}
}

// Mirrors the wire shape the extension emits for near-identical snapshots: a
// full initial snapshot, a delta whose patch flips one rendered flag and
// replaces a text node, and a reference pointing at the delta's sequence
// (the client's exact-match baseline advances past a stored delta).
func deltaRecordingPayload() map[string]any {
	raw := `{
		"version": "2.0.0",
		"snapshots": [
			{
				"timestamp": 1, "url": "https://example.com/", "selectorMap": {"1": {"index": 1, "tagName": "a", "selector": "a.next", "boundingRect": {"x": 0, "y": 0}}},
				"phase": "initial", "sequence": 0, "actionIndex": 0,
				"capture": {"status": "complete", "nodeCount": 10, "frames": []},
				"domTree": {"type": "element", "tagName": "main", "children": [{"type": "text", "text": "Example"}, {"type": "element", "tagName": "section", "rendered": false}]}
			},
			{
				"timestamp": 2, "url": "https://example.com/", "selectorMap": {},
				"phase": "before-action", "sequence": 1, "actionIndex": 0,
				"base": 0,
				"patch": [
					{"op": "replace", "path": "/domTree/children/0/text", "value": "Mutated"},
					{"op": "replace", "path": "/domTree/children/1/rendered", "value": true},
					{"op": "replace", "path": "/selectorMap/1/boundingRect", "value": {"x": 5, "y": 700}}
				]
			},
			{
				"timestamp": 3, "url": "https://example.com/", "selectorMap": {},
				"phase": "before-action", "sequence": 2, "actionIndex": 1, "ref": 1
			}
		]
	}`
	var payload map[string]any
	if err := json.Unmarshal([]byte(raw), &payload); err != nil {
		panic(err)
	}
	return payload
}

func TestExpandSnapshotReferencesAppliesDeltas(t *testing.T) {
	payload := deltaRecordingPayload()
	if err := ExpandSnapshotReferences(payload); err != nil {
		t.Fatalf("expand: %v", err)
	}
	snaps := snapshotsOf(t, payload)
	initial, delta, ref := snaps[0], snaps[1], snaps[2]

	if _, still := delta["patch"]; still {
		t.Fatal("delta kept its patch")
	}
	if _, still := delta["base"]; still {
		t.Fatal("delta kept its base pointer")
	}
	if delta["domTree"] == nil {
		t.Fatal("delta missing expanded domTree")
	}
	tree := delta["domTree"].(map[string]any)
	children := tree["children"].([]any)
	if children[0].(map[string]any)["text"] != "Mutated" {
		t.Fatalf("patched text not applied: %v", children[0])
	}
	if children[1].(map[string]any)["rendered"] != true {
		t.Fatalf("patched rendered flag not applied: %v", children[1])
	}
	rect := delta["selectorMap"].(map[string]any)["1"].(map[string]any)["boundingRect"].(map[string]any)
	if rect["x"] != float64(5) || rect["y"] != float64(700) {
		t.Fatalf("patched boundingRect not applied: %v", rect)
	}
	// The base snapshot itself stays untouched.
	if initial["domTree"].(map[string]any)["children"].([]any)[0].(map[string]any)["text"] != "Example" {
		t.Fatal("base snapshot content was mutated by a delta expansion")
	}
	// Positional identity stays per-snapshot.
	if delta["phase"] != "before-action" || delta["timestamp"] == initial["timestamp"] {
		t.Fatalf("delta lost positional identity: %v", delta)
	}

	// A later reference naming the expanded delta's sequence resolves to it.
	if _, still := ref["ref"]; still {
		t.Fatal("ref kept its pointer")
	}
	if marshal(t, ref["domTree"]) != marshal(t, delta["domTree"]) {
		t.Fatal("ref to expanded delta does not resolve to the patched content")
	}

	// The expanded form satisfies sanitize like any full-snapshot recording.
	sanitized, _ := Sanitize(payload)
	if list, ok := sanitized["snapshots"].([]any); !ok || len(list) != 3 {
		t.Fatalf("sanitized snapshots missing after delta expansion: %v", sanitized["snapshots"])
	}
}

func TestExpandSnapshotDeltaFailsClosed(t *testing.T) {
	mutateDelta := func(edit func(delta map[string]any)) func(map[string]any) {
		return func(p map[string]any) {
			edit(p["snapshots"].([]any)[1].(map[string]any))
		}
	}
	cases := map[string]func(payload map[string]any){
		"unknown base": mutateDelta(func(d map[string]any) { d["base"] = float64(99) }),
		"forward base": mutateDelta(func(d map[string]any) { d["base"] = float64(2) }),
		"negative base": mutateDelta(func(d map[string]any) { d["base"] = float64(-1) }),
		"base without content": mutateDelta(func(d map[string]any) { d["base"] = float64(1) }),
		"patch without base": mutateDelta(func(d map[string]any) { delete(d, "base") }),
		"empty patch": mutateDelta(func(d map[string]any) { d["patch"] = []any{} }),
		"patch not array": mutateDelta(func(d map[string]any) { d["patch"] = "nope" }),
		"unknown op": mutateDelta(func(d map[string]any) {
			d["patch"] = []any{map[string]any{"op": "copy", "path": "/domTree", "value": 1}}
		}),
		"unrooted path": mutateDelta(func(d map[string]any) {
			d["patch"] = []any{map[string]any{"op": "replace", "path": "domTree", "value": 1}}
		}),
		"root path": mutateDelta(func(d map[string]any) {
			d["patch"] = []any{map[string]any{"op": "replace", "path": "/", "value": 1}}
		}),
		"missing path": mutateDelta(func(d map[string]any) {
			d["patch"] = []any{map[string]any{"op": "replace", "path": "/domTree/missing", "value": 1}}
		}),
		"missing value": mutateDelta(func(d map[string]any) {
			d["patch"] = []any{map[string]any{"op": "replace", "path": "/domTree/children/0/text"}}
		}),
		"array index out of bounds": mutateDelta(func(d map[string]any) {
			d["patch"] = []any{map[string]any{"op": "remove", "path": "/domTree/children/9"}}
		}),
		"append token": mutateDelta(func(d map[string]any) {
			d["patch"] = []any{map[string]any{"op": "add", "path": "/domTree/children/-", "value": 1}}
		}),
		"op ceiling": mutateDelta(func(d map[string]any) {
			ops := make([]any, maxSnapshotPatchOps+1)
			for i := range ops {
				ops[i] = map[string]any{"op": "replace", "path": "/domTree/children/0/text", "value": "x"}
			}
			d["patch"] = ops
		}),
		"both ref and patch": mutateDelta(func(d map[string]any) { d["ref"] = float64(0) }),
	}
	for name, mutate := range cases {
		t.Run(name, func(t *testing.T) {
			payload := deltaRecordingPayload()
			mutate(payload)
			err := ExpandSnapshotReferences(payload)
			if !errors.Is(err, ErrInvalidSnapshotDelta) && !errors.Is(err, ErrInvalidSnapshotReference) {
				t.Fatalf("expected a snapshot delta/reference error, got %v", err)
			}
		})
	}
}

func marshal(t *testing.T, value any) string {
	t.Helper()
	encoded, err := json.Marshal(value)
	if err != nil {
		t.Fatal(err)
	}
	return string(encoded)
}

File: src/rule-generator/converter/snapshot-refs.test.ts
import { describe, it, expect } from 'vitest';
import { expandSnapshotReferences } from './snapshot-refs';
import { convert } from './PageAgentToDslConverter';
import type { PageAgentRecording, DomSnapshot } from '../types';

function fullSnapshot(overrides: Partial<DomSnapshot> = {}): DomSnapshot {
  return {
    timestamp: 0,
    url: 'https://example.com/',
    selectorMap: {},
    phase: 'initial',
    sequence: 0,
    domTree: { type: 'element', tagName: 'html', children: [{ type: 'text', text: 'Example' }] },
    capture: { status: 'complete', nodeCount: 2, redactionCount: 0, removedNodeCount: 0, frames: [] },
    ...overrides,
  };
}

function compressedRecording(): PageAgentRecording {
  const initial = fullSnapshot({
    selectorMap: {
      '1': { index: 1, tagName: 'a', selector: 'a.next', boundingRect: { x: 0, y: 0, width: 10, height: 10 } },
    },
  });
  // A delta whose patch registers a NEW element (index 2) in the selector
  // map — exactly the case that broke client-side element resolution.
  const delta = fullSnapshot({
    timestamp: 10,
    phase: 'before-action',
    sequence: 1,
    actionIndex: 0,
    selectorMap: {},
    base: 0,
    patch: [
      { op: 'add', path: '/selectorMap/2', value: { index: 2, tagName: 'button', selector: 'button.go', boundingRect: { x: 5, y: 5, width: 8, height: 8 } } },
      { op: 'replace', path: '/domTree/children/0/text', value: 'Mutated' },
    ],
  });
  // A later identical capture referencing the DELTA's sequence.
  const ref = fullSnapshot({
    timestamp: 20,
    phase: 'final',
    sequence: 2,
    selectorMap: {},
    ref: 1,
  });
  return {
    version: '2.0.0',
    meta: {
      startUrl: 'https://example.com/',
      title: 'Example',
      recordedAt: '2026-09-24T08:00:00Z',
      endedAt: '2026-09-24T08:01:00Z',
      domain: 'example.com',
      semanticDomVersion: '1',
      sanitizationVersion: 'extension-v2',
    },
    limits: { maxActions: 500, maxDurationMs: 7200000, maxBytes: 26214400, warningThreshold: 0.8 },
    warnings: [],
    termination: { reason: 'user', message: 'done', timestamp: 25, complete: true },
    events: [{ type: 'click', index: 2, timestamp: 10 }],
    snapshots: [initial, delta, ref],
  };
}

describe('expandSnapshotReferences (client-side)', () => {
  it('expands deltas, allows refs to name expanded deltas, and keeps positional identity', () => {
    const recording = compressedRecording();
    expandSnapshotReferences(recording);
    const [initial, delta, ref] = recording.snapshots;

    expect(delta.patch).toBeUndefined();
    expect(delta.base).toBeUndefined();
    expect(delta.domTree).toBeDefined();
    expect((delta.domTree as { children: Array<{ text?: string }> }).children[0].text).toBe('Mutated');
    expect(delta.selectorMap?.['2']?.selector).toBe('button.go');
    expect(delta.phase).toBe('before-action');
    expect(delta.timestamp).toBe(10);

    // The base snapshot itself stays untouched.
    expect((initial.domTree as { children: Array<{ text?: string }> }).children[0].text).toBe('Example');
    expect(initial.selectorMap?.['2']).toBeUndefined();

    // The ref resolved to the expanded delta's content.
    expect(ref.ref).toBeUndefined();
    expect(JSON.stringify(ref.domTree)).toBe(JSON.stringify(delta.domTree));
    expect(ref.selectorMap?.['2']?.selector).toBe('button.go');
  });

  it('lets the rule converter resolve an element whose index only exists in a delta patch', () => {
    // This is the exact regression: without expansion, convert() threw
    // "Element not found in recording snapshots: index=2".
    const recording = compressedRecording();
    expandSnapshotReferences(recording);
    const rule = convert(recording, { ruleIdPrefix: 'ext' });
    // The click resolved to a real target (the optimizer may rewrite the
    // raw selector string, e.g. button.go → .go).
    const step = rule.steps.find((s) => s.action === 'click') as { target: { $ref?: string } };
    expect(step).toBeDefined();
    expect(step.target?.$ref).toBeDefined();
    const alias = rule.selectors?.[step.target.$ref as string] as { selector?: string } | undefined;
    expect(alias?.selector).toBeTruthy();
    expect(alias?.selector).toContain('go');
  });

  it('fails closed on malformed pointers, patches, and unknown bases', () => {
    const cases: Array<{ name: string; mutate: (rec: PageAgentRecording) => void }> = [
      { name: 'unknown ref', mutate: (r) => { r.snapshots[2].ref = 99; } },
      { name: 'forward ref', mutate: (r) => { r.snapshots[0].ref = 2; } },
      { name: 'negative ref', mutate: (r) => { r.snapshots[2].ref = -1; } },
      { name: 'ref and patch together', mutate: (r) => { (r.snapshots[1] as DomSnapshot & { ref?: number }).ref = 0; } },
      { name: 'unknown base', mutate: (r) => { r.snapshots[1].base = 42; } },
      { name: 'bad patch path', mutate: (r) => { r.snapshots[1].patch = [{ op: 'replace', path: 'domTree', value: 1 }]; } },
      { name: 'valueless op', mutate: (r) => { r.snapshots[1].patch = [{ op: 'replace', path: '/domTree' }]; } },
      { name: 'missing path', mutate: (r) => { r.snapshots[1].patch = [{ op: 'replace', path: '/domTree/missing', value: 1 }]; } },
    ];
    for (const { name, mutate } of cases) {
      const recording = compressedRecording();
      mutate(recording);
      expect(() => expandSnapshotReferences(recording), name).toThrow();
    }
  });

  it('is a no-op for legacy v1 recordings and snapshot-less payloads', () => {
    const legacy = { version: '1.0.0', snapshots: [{ timestamp: 1, url: 'https://example.com/', selectorMap: {} }] } as unknown as PageAgentRecording;
    expect(() => expandSnapshotReferences(legacy)).not.toThrow();
  });
});
